c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -223,7 +231,325 @@ struct Block {
+(cuMemCreate) from the allocation of virtual address space (cuMemAddressReserve)
+and the associate between them cuMemMap/cuMemSetAccess.
+
+When we allocate a new segment, we allocate enough address space to map
+basically the entire physical memory of the GPU (there is 256TiB of address
+space), but we only map enough physical memory to handle the current amount of
+memory needed by the program. As more is requested, we add more physical memory
+to the segment. This can work at the granularity of GPU pages which are 2MiB
+currently.
+
+If we end up out of memory, we can unmap all the memory in our segmenet
+corresponding to empty physical pages, and return it to CUDA for use at another
+address in the segment or in a segment for a different stream.
+
+A current limitation of CUDA's API is that physical memory cannot be split up
+after it is mapped, and the cost to map/unmap memory is proportional the number
+of physical memory chunks that were allocated (mapping 10 separately allocated
+2MiB pages takes 10x time compared to mapping one 20MiB physical allocation of
+10 pages).  So we use 2MiB pages for our small pool and 20MiB pages for our
+large pool. Initially allocation using expandable_blocks will be slower than
+cudaMalloc, though still in the milliseconds range for mapping the entire
+memory.
+
+When mapping new memory to expand the segment, we look for the lowest address at
+which we can fit a new allocation by adding new pages. Normally this will be at
+the end of the block. But if have previously unmapped blocks earlier in the
+segment during an OOM, it will first try to fill in those gaps to keep the
+segment as a single block.
+
+Allocation of blocks in the segment uses the same best-fit heuristics of the
+rest of the allocator.
+
+Expandable blocks can be enabled/disabled throughout the run of a program. When
+disabled, the allocator will not put new allocations in an expandable block.
+
+Limitations
+
+* Slightly slower initial memory allocation speed.
+* IPC of cuda tensors (e.g. for multiprocess dataloaders) is not supported.
+However, it is possible to temporarily disable (expandable_segments:False) the
+bevhavior for allocator tensors that need to be used cross-process.
+* CUDA runtime APIs related to sharing memory across process
+(cudaDeviceEnablePeerAccess) do not work for memory allocated with cuMemMap.
+Instead these mapping have to be done manually. The allocator now has an
+`enablePeerAccess` method to do this.
+*/
+
+struct ExpandableSegment {
+  ExpandableSegment(
+      int device,
+      cudaStream_t stream,
+      size_t size,
+      const std::vector<int>& peers)
+      : device_(device),
+        stream_(stream),
+        max_handles_(0),
+        segment_size_(size),
+        peers_(peers) {
+    cudaDeviceProp prop;
+    C10_CUDA_CHECK(cudaGetDeviceProperties(&prop, device_));
+    max_handles_ = numSegments(prop.totalGlobalMem + prop.totalGlobalMem / 8);
```

Comment:
8 is number of devices ?

Reply:
It's an extra factor beyond the physical memory of the current device to allow the segment to accurate some unmapped pages inside of it before we have to allocate a whole new segment. I'm not sure if it is really needed but it seems likely that after an OOM we might end up with some blank internal pages that never get reused well.

Reply:
add comment ?